File: Backend/app/schemas/listings.py
from pydantic import BaseModel, Field, field_validator, computed_field
from typing import Optional, List
from decimal import Decimal
from datetime import datetime
import uuid

# Import media response for including images in listing
from typing import TYPE_CHECKING
if TYPE_CHECKING:
    from app.schemas.media import ListingMediaResponse

class CreateListingRequest(BaseModel):
    title: str = Field(..., min_length=2, max_length=500)
    description: Optional[str] = None
    operation: str = Field(..., description="Operation type: rent, sale, temp_rent")
    property_type: str = Field(..., description="Property type: apartment, house, office, land, etc.")
    price: Decimal = Field(..., ge=0)
    currency: Optional[str] = "PEN"
    
    # Area information
    area_built: Optional[Decimal] = None
    area_total: Optional[Decimal] = None
    
    # Property details
    bedrooms: Optional[int] = None
    bathrooms: Optional[int] = None
    parking_spots: Optional[int] = None
    floors: Optional[int] = None
    floor_number: Optional[int] = None
    age_years: Optional[int] = None
    pet_friendly: Optional[bool] = Field(None, description="Whether the property allows pets")
    furnished: Optional[bool] = Field(None, description="Whether the property is furnished")
    rental_mode: Optional[str] = Field(None, description="Rental mode: full_property, private_room, shared_room")
    rental_term: Optional[str] = Field(None, description="Rental term: daily, weekly, monthly, yearly")
    rental_model: Optional[str] = Field('traditional', description="Rental model: traditional or airbnb")
    airbnb_score: Optional[int] = Field(None, ge=0, le=100, description="Airbnb eligibility score (0-100)")
    airbnb_eligible: Optional[bool] = Field(None, description="Whether the property is eligible for Airbnb")
    airbnb_opted_out: Optional[bool] = Field(False, description="Whether the owner opted out of Airbnb functionality")
    
    # Location
    address: Optional[str] = None
    department: Optional[str] = None
    province: Optional[str] = None
    district: Optional[str] = None
    latitude: Optional[Decimal] = None
    longitude: Optional[Decimal] = None
    
    # Airbnb-specific fields
    smoking_allowed: Optional[bool] = Field(None, description="Whether smoking is allowed")
    deposit_required: Optional[bool] = Field(False, description="Whether deposit is required")
    deposit_amount: Optional[Decimal] = Field(None, ge=0, description="Deposit amount if required")
    minimum_stay_nights: Optional[int] = Field(1, ge=1, description="Minimum stay in nights")
    maximum_stay_nights: Optional[int] = Field(None, ge=1, description="Maximum stay in nights")
    check_in_time: Optional[str] = Field(None, description="Check-in time (HH:MM format)")
    check_out_time: Optional[str] = Field(None, description="Check-out time (HH:MM format)")
    max_guests: Optional[int] = Field(None, ge=1, description="Maximum number of guests")
    cleaning_included: Optional[bool] = Field(False, description="Whether cleaning is included")
    cleaning_fee: Optional[Decimal] = Field(None, ge=0, description="Cleaning fee amount")
    utilities_included: Optional[bool] = Field(False, description="Whether utilities are included")
    internet_included: Optional[bool] = Field(False, description="Whether internet is included")
    house_rules: Optional[str] = Field(None, description="House rules text")
    cancellation_policy: Optional[str] = Field('flexible', description="Cancellation policy: flexible, moderate, strict")
    available_from: Optional[str] = Field(None, description="Date when property is available (YYYY-MM-DD)")
    
    # Amenities
    amenities: Optional[List[int]] = Field(default=[], description="List of amenity IDs")
    
    # Contact information
    contact_name: Optional[str] = Field(None, description="Contact person name")
    contact_phone_e164: Optional[str] = Field(None, description="Contact phone in E.164 format")
    contact_whatsapp_phone_e164: Optional[str] = Field(None, description="WhatsApp phone in E.164 format")
    contact_whatsapp_link: Optional[str] = Field(None, description="Pre-generated WhatsApp link")

class UpdateListingRequest(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    price: Optional[Decimal] = None
    currency: Optional[str] = None
    area_built: Optional[Decimal] = None
    area_total: Optional[Decimal] = None
    bedrooms: Optional[int] = None
    bathrooms: Optional[int] = None
    parking_spots: Optional[int] = None
    floors: Optional[int] = None
    floor_number: Optional[int] = None
    age_years: Optional[int] = None
    pet_friendly: Optional[bool] = None
    furnished: Optional[bool] = None
    rental_mode: Optional[str] = None
    rental_term: Optional[str] = None
    rental_model: Optional[str] = None
    airbnb_score: Optional[int] = None
    airbnb_eligible: Optional[bool] = None
    airbnb_opted_out: Optional[bool] = None
    address: Optional[str] = None
    department: Optional[str] = None
    province: Optional[str] = None
    district: Optional[str] = None
    latitude: Optional[Decimal] = None
    longitude: Optional[Decimal] = None
    
    # Airbnb-specific fields
    smoking_allowed: Optional[bool] = None
    deposit_required: Optional[bool] = None
    deposit_amount: Optional[Decimal] = None
    minimum_stay_nights: Optional[int] = None
    maximum_stay_nights: Optional[int] = None
    check_in_time: Optional[str] = None
    check_out_time: Optional[str] = None
    max_guests: Optional[int] = None
    cleaning_included: Optional[bool] = None
    cleaning_fee: Optional[Decimal] = None
    utilities_included: Optional[bool] = None
    internet_included: Optional[bool] = None
    house_rules: Optional[str] = None
    cancellation_policy: Optional[str] = None
    available_from: Optional[str] = None
    amenities: Optional[List[int]] = None
    
    # Contact information
    contact_name: Optional[str] = None
    contact_phone_e164: Optional[str] = None
    contact_whatsapp_phone_e164: Optional[str] = None
    contact_whatsapp_link: Optional[str] = None

class ListingResponse(BaseModel):
    id: str
    title: str
    description: Optional[str]
    operation: str
    property_type: str
    price: Decimal
    currency: str
    area_built: Optional[Decimal]
    area_total: Optional[Decimal]
    bedrooms: Optional[int]
    bathrooms: Optional[int]
    parking_spots: Optional[int]
    pet_friendly: Optional[bool]
    furnished: Optional[bool]
    rental_mode: Optional[str]
    rental_term: Optional[str]
    rental_model: Optional[str]
    airbnb_score: Optional[int]
    airbnb_eligible: Optional[bool]
    airbnb_opted_out: Optional[bool]
    
    # Airbnb-specific fields
    smoking_allowed: Optional[bool]
    deposit_required: Optional[bool]
    deposit_amount: Optional[Decimal]
    minimum_stay_nights: Optional[int]
    maximum_stay_nights: Optional[int]
    check_in_time: Optional[str]
    check_out_time: Optional[str]
    max_guests: Optional[int]
    cleaning_included: Optional[bool]
    cleaning_fee: Optional[Decimal]
    utilities_included: Optional[bool]
    internet_included: Optional[bool]
    house_rules: Optional[str]
    cancellation_policy: Optional[str]
    available_from: Optional[str]
    
    @computed_field
    @property
    def is_airbnb_available(self) -> Optional[bool]:
        """Computed field: whether property is available for Airbnb"""
        if self.airbnb_eligible is None:
            return None
        return (
            bool(self.airbnb_eligible) and 
            not bool(self.airbnb_opted_out) and 
            self.operation in ['rent', 'temp_rent']
        )
    address: Optional[str]
    department: Optional[str]
    province: Optional[str]
    district: Optional[str]
    latitude: Optional[Decimal]
    longitude: Optional[Decimal]
    
    # Contact information
    contact_name: Optional[str]
    contact_phone_e164: Optional[str]
    contact_whatsapp_phone_e164: Optional[str]
    contact_whatsapp_link: Optional[str]
    
    status: str
    verification_status: str
    owner_user_id: str
    agency_id: Optional[str]
    views_count: int
    leads_count: int
    favorites_count: int
    has_media: bool
    images: List[dict] = []  # Lista de imágenes
    created_at: datetime
    updated_at: datetime
    published_at: Optional[datetime]

    @field_validator('id', 'owner_user_id', 'agency_id', mode='before')
    @classmethod
    def convert_uuid_to_str(cls, v):
        if isinstance(v, uuid.UUID):
            return str(v)
        return v
    
    @field_validator('check_in_time', 'check_out_time', mode='before')
    @classmethod
    def convert_time_to_str(cls, v):
        if v is None:
            return v
        if hasattr(v, 'isoformat'):
            return v.isoformat()
        return str(v) if v else None
    
    @field_validator('available_from', mode='before')
    @classmethod
    def convert_date_to_str(cls, v):
        if v is None:
            return v
        if hasattr(v, 'isoformat'):
            return v.isoformat()
        return str(v) if v else None

    class Config:
        from_attributes = True

class ChangeListingStatusRequest(BaseModel):
    status: str = Field(..., description="New status: draft, published, unpublished, archived")

class RegisterViewRequest(BaseModel):
    viewer_id: Optional[str] = None
    metadata: Optional[dict] = None

File: Frontend/web/lib/api/listings.ts
// API functions for listings/properties
import { API_BASE_URL, authenticatedRequest } from './auth';

export interface Listing {
  id: string;
  title: string;
  description?: string;
  operation: string; // rent, sale, temp_rent
  property_type: string;
  price: number;
  currency: string;
  
  // Area information
  area_built?: number;
  area_total?: number;
  
  // Property details
  bedrooms?: number;
  bathrooms?: number;
  parking_spots?: number;
  floors?: number;
  floor_number?: number;
  age_years?: number;
  pet_friendly?: boolean;
  furnished?: boolean;
  rental_mode?: string; // full_property, private_room, shared_room
  rental_term?: string; // daily, weekly, monthly, yearly
  rental_model?: string; // traditional or airbnb
  airbnb_score?: number;
  airbnb_eligible?: boolean;
  airbnb_opted_out?: boolean;
  is_airbnb_available?: boolean;
  
  // Location
  address?: string;
  department?: string;
  province?: string;
  district?: string;
  latitude?: number;
  longitude?: number;
  
  // Airbnb-specific fields
  smoking_allowed?: boolean;
  deposit_required?: boolean;
  deposit_amount?: number;
  minimum_stay_nights?: number;
  maximum_stay_nights?: number;
  check_in_time?: string;
  check_out_time?: string;
  max_guests?: number;
  cleaning_included?: boolean;
  cleaning_fee?: number;
  utilities_included?: boolean;
  internet_included?: boolean;
  house_rules?: string;
  cancellation_policy?: string;
  available_from?: string;
  
  // Contact information
  contact_name?: string;
  contact_phone_e164?: string;
  contact_whatsapp_phone_e164?: string;
  contact_whatsapp_link?: string;
  
  // Media
  images?: ListingImage[];
  amenities?: Array<{ id: number; name: string; icon?: string }>;
  
  // Status and metadata
  status: string;
  verification_status: string;
  owner_user_id: string;
  agency_id?: string;
  views_count: number;
  leads_count: number;
  favorites_count: number;
  has_media: boolean;
  created_at: string;
  updated_at: string;
  published_at?: string;
}

export interface ListingImage {
  id: string;
  url: string;
  caption?: string;
  display_order: number;
  is_main: boolean;
}

export interface CreateListingRequest {
  title: string;
  description?: string;
  property_type: string;
  rental_type: string;
  rental_term?: string;
  price_amount: number;
  price_currency?: string;
  address_line1: string;
  address_line2?: string;
  city: string;
  state_province?: string;
  postal_code?: string;
  country?: string;
  latitude?: number;
  longitude?: number;
  bedrooms?: number;
  bathrooms?: number;
  area_size?: number;
  area_unit?: string;
  status?: string;
}

/**
 * Get all listings for the current user
 */
export async function getMyListings(status?: string): Promise<Listing[]> {
  try {
    const url = status 
      ? `${API_BASE_URL}/v1/listings/my?status=${status}`
      : `${API_BASE_URL}/v1/listings/my`;

    const response = await authenticatedRequest(url, {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
        'Accept': 'application/json',
      },
    });

    if (!response.ok) {
      const error = await response.json().catch(() => ({}));
      throw new Error(error.detail || `Error fetching listings: ${response.status}`);
    }

    return await response.json();
  } catch (error) {
    console.error('Error fetching my listings:', error);
    throw error;
  }
}

/**
 * Get a single listing by ID
 */
export async function getListing(listingId: string): Promise<Listing> {
  try {
    const response = await authenticatedRequest(`${API_BASE_URL}/v1/listings/${listingId}`, {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
        'Accept': 'application/json',
      },
    });

    if (!response.ok) {
      const error = await response.json().catch(() => ({}));
      throw new Error(error.detail || `Error fetching listing: ${response.status}`);
    }

    return await response.json();
  } catch (error) {
    console.error('Error fetching listing:', error);
    throw error;
  }
}

/**
 * Create a new listing
 */
export async function createListing(data: CreateListingRequest): Promise<Listing> {
  try {
    const response = await authenticatedRequest(`${API_BASE_URL}/v1/listings/`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Accept': 'application/json',
      },
      body: JSON.stringify(data),
    });

    if (!response.ok) {
      const error = await response.json().catch(() => ({}));
      throw new Error(error.detail || `Error creating listing: ${response.status}`);
    }

    return await response.json();
  } catch (error) {
    console.error('Error creating listing:', error);
    throw error;
  }
}

/**
 * Update an existing listing
 */
export async function updateListing(listingId: string, data: Partial<CreateListingRequest>): Promise<Listing> {
  try {
    const response = await authenticatedRequest(`${API_BASE_URL}/v1/listings/${listingId}`, {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
        'Accept': 'application/json',
      },
      body: JSON.stringify(data),
    });

    if (!response.ok) {
      const error = await response.json().catch(() => ({}));
      throw new Error(error.detail || `Error updating listing: ${response.status}`);
    }

    return await response.json();
  } catch (error) {
    console.error('Error updating listing:', error);
    throw error;
  }
}

/**
 * Delete a listing
 */
export async function deleteListing(listingId: string): Promise<void> {
  try {
    const response = await authenticatedRequest(`${API_BASE_URL}/v1/listings/${listingId}`, {
      method: 'DELETE',
      headers: {
        'Accept': 'application/json',
      },
    });

    if (!response.ok) {
      const error = await response.json().catch(() => ({}));
      throw new Error(error.detail || `Error deleting listing: ${response.status}`);
    }
  } catch (error) {
    console.error('Error deleting listing:', error);
    throw error;
  }
}

/**
 * Publish a listing
 */
export async function publishListing(listingId: string): Promise<Listing> {
  try {
    const response = await authenticatedRequest(`${API_BASE_URL}/v1/listings/${listingId}/publish`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Accept': 'application/json',
      },
    });

    if (!response.ok) {
      const error = await response.json().catch(() => ({}));
      throw new Error(error.detail || `Error publishing listing: ${response.status}`);
    }

    return await response.json();
  } catch (error) {
    console.error('Error publishing listing:', error);
    throw error;
  }
}

/**
 * Unpublish a listing
 */
export async function unpublishListing(listingId: string): Promise<Listing> {
  try {
    const response = await authenticatedRequest(`${API_BASE_URL}/v1/listings/${listingId}/unpublish`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Accept': 'application/json',
      },
    });

    if (!response.ok) {
      const error = await response.json().catch(() => ({}));
      throw new Error(error.detail || `Error unpublishing listing: ${response.status}`);
    }

    return await response.json();
  } catch (error) {
    console.error('Error unpublishing listing:', error);
    throw error;
  }
}

/**
 * Get images for a listing
 */
export async function getListingImages(listingId: string): Promise<ListingImage[]> {
  try {
    const response = await authenticatedRequest(`${API_BASE_URL}/v1/media/listings/${listingId}/images`, {
      method: 'GET',
      headers: {
        'Accept': 'application/json',
      },
    });

    if (!response.ok) {
      const error = await response.json().catch(() => ({}));
      throw new Error(error.detail || `Error fetching images: ${response.status}`);
    }

    const data = await response.json();
    return data.images || [];
  } catch (error) {
    console.error('Error fetching listing images:', error);
    throw error;
  }
}
